pkg/azuredisk/nodeserver_v2.go

```diff
@@ -519,3 +537,78 @@ func (d *DriverV2) ensureMountPoint(target string) (bool, error) {
 	klog.Warning("ensureMountPoint method is deprecated.")
 	return d.nodeProvisioner.EnsureMountPointReady(target)
 }
+
+func (d *DriverV2) recoverMount(diskURI string) {
+	// if mount recovery is necessary (if time.Now is past the set nextRetry), detach volume so that volume can be reattached via reconciliation
+	if d.shouldStartRecovery(diskURI) {
+		klog.Warningf("Starting mount recovery: detaching the volume (%s) from node (%s)...", diskURI, d.NodeID)
+		recoveryCtx := context.Background()
+
+		// When the device path cannot be found for the specified LUN, try to recover by detaching and re-attaching the disk.
+		// We only need to perform the detach. The CSI infrastructure will re-attach the disk when it notices a difference in
+		// world states (ListVolumes result doesn't match VolumeAttachment list).
+		detachErr := d.crdProvisioner.UnpublishVolume(recoveryCtx, diskURI, d.NodeID, nil)
+		if detachErr != nil {
+			klog.Errorf("failed to unpublishVolume volume (%s) from node (%s): %v", diskURI, d.NodeID, detachErr)
+			d.markDetachState(diskURI, detachFailed)
+		} else {
+			klog.Infof("Detached volume (%s) from node (%s) successfully...", diskURI, d.NodeID)
+			d.markDetachState(diskURI, detachCompleted)
+		}
+	}
+}
+
+func (d *DriverV2) shouldStartRecovery(diskURI string) bool {
+	d.deviceChecker.lock.Lock()
+	defer d.deviceChecker.lock.Unlock()
+
+	recoveryInProcess, diskURIMatches := d.isRecoveryInProcess(diskURI)
+	if !recoveryInProcess {
+		d.deviceChecker.entry = newDeviceCheckerEntry(diskURI)
+		return true
+	}
+	// if there already is another recovery in process for different volume, no need to start recovery because recovery for that other volume will likely fix the problem for this volume
+	return diskURIMatches
+}
+
+// isRecoveryInProcess assumes that lock has been acquire and should be used only AFTER ACQUIRING THE LOCK
+func (d *DriverV2) isRecoveryInProcess(diskURI string) (recoveryInProcess, diskURIMatches bool) {
+	if d.deviceChecker.entry != nil {
+		recoveryInProcess = true
+		diskURIMatches = d.deviceChecker.entry.diskURI == diskURI
+	}
+	return
+}
+
+func (d *DriverV2) getDetachState(diskURI string) int32 {
+	d.deviceChecker.lock.RLock()
+	defer d.deviceChecker.lock.RUnlock()
+
+	if recoveryInProcess, diskURIMatches := d.isRecoveryInProcess(diskURI); !recoveryInProcess || !diskURIMatches {
+		return detachNotFound
+	}
+
+	return atomic.LoadInt32(&d.deviceChecker.entry.detachState)
+}
+
+func (d *DriverV2) markDetachState(diskURI string, detachState int32) {
+	d.deviceChecker.lock.RLock()
+	defer d.deviceChecker.lock.RUnlock()
+
+	if recoveryInProcess, diskURIMatches := d.isRecoveryInProcess(diskURI); !recoveryInProcess || !diskURIMatches {
+		return
+	}
+
+	atomic.StoreInt32(&d.deviceChecker.entry.detachState, detachState)
```

Comment:
This method is expected to modify state, so acquiring the exclusive lock is more intuitive. This would also allow you to read and update `detachState` w/o relying on atomic functions.